core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -931,11 +931,15 @@ private[log] class Cleaner(val id: Int,
             stats.indexMessagesRead(batch.countOrNull)
           } else {
             for (record <- batch.asScala) {
-              if (record.hasKey && record.offset >= startOffset) {
-                if (map.size < maxDesiredMapSize)
-                  map.put(record.key, record.offset)
-                else
-                  return true
+              if (record.hasKey) {
+                if (record.offset >= startOffset) {
+                  if (map.size < maxDesiredMapSize)
+                    map.put(record.key, record.offset)
+                  else
+                    return true
+                } else {
+                  map.updateLatestOffset(record.offset + 1)
```

Comment:
I am not sure if this completely fixes the issue. Let's say we have one non active segment with offset range [100, 190) and an active segment of [200, 300). With this change, we will set the latestOffset in the map to 190 (which is an improvement from 0). However, this can still trigger another round of unnecessary cleaning since the first segment will still be treated as cleanable (instead of cleaned).

One way to fix this more completely is that when we have finished iterating all batches in a segment, we update the latest offset in the map to the start offset of the next segment. In the above example, this will set the latest offset of the map to 200 and the next round of cleaning won't be triggered until the active segment rolls.

Reply:
You were right here. Please, have a look at new fix proposal, and test covering specific issue.